project/TestVariant.scala

```diff
@@ -0,0 +1,150 @@
+        }
+
+        object State {
+            val zero: State = State(Vector.empty, Mode.Continue)
+        }
+    }
+
+    val generate = Def.task {
+
+        import internal.*
+
+        implicit val log: ManagedLogger = streams.value.log
+        val files: Seq[File]            = (Test / unmanagedSources).value
+        val outDir                      = (Test / sourceManaged).value / "testVariants"
+
+        lazy val created: Boolean = {
+            log.debug(s"Creating directory $outDir")
+            IO.createDirectory(outDir)
+            outDir.isDirectory
+        }
+
+        var fileCounter = 0
+
+        def newScalaFile(suffix: String): File = {
+            require(created)
+            val nn: String = if (fileCounter < 10) "0" + fileCounter else "" + fileCounter
+            val file       = outDir / s"generated$nn-$suffix.scala"
+            fileCounter = fileCounter + 1
+
+            log.debug(s"Creating file $file")
+            file.createNewFile()
+            file
+        }
+
+        files.flatMap(file => {
+            val content = IO.read(file, StandardCharsets.UTF_8)
+
+            if (content.contains("@TestVariant")) {
+                log.debug(s"Processing file $file for variants")
+
+                var expecedSize: Option[Int] = None
+
+                val processed: State = content.split("\n").foldLeft(State.zero)((state, str) => {
+
+                    val nextMod: Mode = Variant.extractFrom(str) match {
+                        case Some(Success(variant)) if expecedSize.isEmpty =>
+                            expecedSize = Some(variant.replacements.size)
+                            Mode.Replace(variant)
+
+                        case Some(Success(variant)) if expecedSize.contains(variant.replacements.size) =>
+                            Mode.Replace(variant)
+
+                        case Some(Success(variant)) =>
+                            log.error(s"in $file\ninvalid variant $str : ${expecedSize.map(i =>
+                                    s"expected size $i, actual size ${variant.replacements.size}"
+                                ).mkString}")
+                            throw sys.error(s"invalid variant $str for $file")
+
+                        case Some(Failure(exception)) =>
+                            log.error(s"cannot generate variant for $file")
+                            throw exception
+                        case None => Mode.Continue
+                    }
+
+                    val line: Line = state.mode match {
+                        case Mode.Continue => Line.Raw(str)
+                        case Mode.Replace(testVariant) =>
+                            if (str.contains(testVariant.base))
+                                Line.Variants(testVariant.replacements.map(r => str.replace(testVariant.base, r)))
+                            else {
+                                log.error(s"in $file: \ncannot find '${testVariant.base}' in this line: \n|  $str")
```

Comment:
Nice-to-have, since we have the file, it should be possible to log `$path:$lineNumber` so that the user can navigate to the file and line in the IDE. I think this should just require `content.split(...).zipWithIndex.foldLeft(...`?